editor/editor_file_system.cpp

```diff
@@ -2620,7 +2620,10 @@ bool EditorFileSystem::_scan_extensions() {
 	Vector<String> loaded_extensions = GDExtensionManager::get_singleton()->get_loaded_extensions();
 	for (int i = 0; i < loaded_extensions.size(); i++) {
 		if (!extensions.has(loaded_extensions[i])) {
-			extensions_removed.push_back(loaded_extensions[i]);
+			// The extension may not have a .gdextension file.
+			if (!FileAccess::exists(loaded_extensions[i])) {
+				extensions_removed.push_back(loaded_extensions[i]);
+			}
```

Comment:
While this will work in many cases, it'll have problems if the loader system is used for "embedded" GDExtensions (like with PR https://github.com/godotengine/godot/pull/90510), which won't have any file on the filesystem.

I wonder if we should be somehow asking the loader if the extension is still available?

Reply:
I didn't think of embedded GDExtensions. I've been assuming extensions always have a file (whether that file is a `.gdextension` or a `.dll`).

Since the loader should know the path to the file(s) it needs to load the extension, I guess it makes sense to let it handle this.

Reply:
Actually, now that I think about it. Since `GDExtensionManager::load_extension_with_loader` takes a `path` parameter. What does this `path` contain for embedded GDExtensions? An empty string?

And since the `GDExtension` class is a resource, and resources should always have a path. What path should be set for these types of extensions?

https://github.com/godotengine/godot/blob/c4560fff5bf86273dd0648a759caa450a6e92183/core/extension/gdextension_manager.cpp#L97

Reply:
It will be some fake path. The [current LibGodot PR](https://github.com/godotengine/godot/pull/90510) uses `"res://__LibGodot"`, but it's also only allowing a single embedded extension at the moment.

It may make sense to use a new prefix like `"libgodot://"` and then each embedded extension could give some unique name like `"libgodot://extension1"`, `"libgodot://extension2"`, etc?

Anyway, that's one of the things that still needs to get figured out :-)

Reply:
How would it work if there is only a `.dll`? What about specifying things like the minimum engine version, the icons, or the paths to files for other platforms/architectures? I'm confused as how we could not have a `.gdextension` file.

Reply:
The "only a `.dll` case" is for a special .NET GDExtension, which is built from the C# code for the current project. The loader for it would be in a Godot module. This is part of the project moving .NET to GDExtension that Raul is working on.

In the LibGodot case, there is no `.gdextension` file because the GDExtension is part of the app that's embedding Godot.

> What about specifying things like the minimum engine version, the icons, or the paths to files for other platforms/architectures?

No one is really happy with the icons being the `.gdextension` file - we need to come up with an alternative for that eventually, anyway. :-)

In the LibGodot case, the "minimum engine version" doesn't matter, since you're embedding a specific version of Godot. Same for the platforms/architectures, since both Godot and the GDExtension are built together.

In the .NET case, I'm not sure how the "minimum engine version" comes into play. But, for the platforms/architectures, I think this is dealing with the portable .NET `.dll`s that contain IL, rather than machine code.

Reply:
> In the .NET case, I'm not sure how the "minimum engine version" comes into play. 

The version of the .NET bindings that will be used by the user project will be determined by the `Godot.NET.Sdk` and this strictly matches the version of Godot that's currently executing. Same as the current behavior with the C# Scripting API.

> But, for the platforms/architectures, I think this is dealing with the portable .NET `.dll`s that contain IL, rather than machine code.

That's correct, it's a portable .NET `.dll` that contains IL. Native dependencies are loaded by the usual .NET mechanism.

Reply:
I just got around to checking out this PR in more detail.
I have a couple notes:
  * I think it would be beneficial to allow the definition of GDExtensionLoaders inside GDExtensions, similar to what I proposed here: https://docs.google.com/document/d/1QwZpo3oIKHyita1mDOIB_xIxdwJ5rtAmpY4_vP9S02Y/edit#heading=h.lltwuthuzzcc (Please note that the implementation details in this old proposal are pretty simplistic and the solution I describe here in this is much more versatile.)
    * This would mean that Godot needs to have only 2 Loader implementations in core: the current shared library / dlopen based one, and the LibGodot one (that is using just a function pointer to an initialization function)
    * The .NET GDExtension implementation could implement the .NET specific loader in itself
    * The same way other languages could provide their own loaders (e.g. .jar loader for Java / Kotlin people, or direct scripting language loaders e.g. for Python, JavaScript ... etc. language support)
  * I think that we should take this opportunity to move away from the path based GDExtension identification concept to a more generic solution. For each GDExtension we could have a GDExtensionDescriptor (which is RefCounted), that uniquely describes that GDExtension. Each GDExtensionLoader subclass could also define one or more GDExtensionDescriptor subclasses that it can handle.
    * For the current library loader this would mean the specification of file paths, symbol names ... etc.
    * For the LibGodot loader, it just means a function pointer

For the GDExtensionManager we would add 2 methods:
  * ``register_loader(String loader_id, Ref<GDExtensionLoader> loader)``
  * ``load_extension(Ref<GDExtensionDescriptor> def)``

These would handle registering loaders and loading extensions.
    
How should we handle the current .gdextension files?
  * My proposal is to extend it with a loader="loader_id" key.
  * On startup the library and (if enabled) the libgodot loaders are registered. For backward compatibility, if no loader= key is found, the library loader will be assumed as default.
  * During startup, each GDExtension file found in the project is parsed into a ``Dictionary<String, String>`` or similar suitable data structure. these would be stored in a separate list for each loader. (the full data structure would be probably something like: ``Dictionary<String, Array<Dictionary<String, String>>>``)
  * The GDExtensionLoader interface would be extended with the following method: ``Ref<GDExtensionDescriptor> create_descriptor_from_dict(Dictionary file_contents)``
    * This function will return a reference to the appropriate GDExtensionDescriptor subclass, that the loader will be able to handle. 
  * For some extensions, where the gdextension file based definition does not make sense (e.g. the LibGodot loader) this function will always return an Error / invalid reference.
  * After the parsing of the gdextension files is completed, the following pseudocode could handle the loading of the defined extensions:
  
```
// The parsed gdextension files
gdextension_data: Dictionary<String, Array<Dictionary<String, String>>> = {...}
// The list of loaders already defined
list_of_loaders = [...] 
while (loader = list_of_loaders.pop()) {
    for (extension_def in gdextension_data[loader]) {
        extension_descriptor = loader.create_descriptor_from_dict(extension_def)
        // The following call could potentially define a new loader and extend the list_of_loaders list.
        loader.load_extension(extension_descriptor) 
    }
}
```

What do you think?

Reply:
> I think it would be beneficial to allow the definition of GDExtensionLoaders inside GDExtensions

This is a very interesting idea, and I think it's something we should allow eventually! But I don't think we should rush into it.

The way the current PR is written has no effect on Godot's external API - the `GDExtensionLoader` here is a completely internal concept. This allows us to try it out, and iterate on it, without the risk of accidentally committing to a broken API that we need to maintain compatibility for.

> The .NET GDExtension implementation could implement the .NET specific loader in itself

Raul will need to speak to this, but it was my understanding that the current plan for .NET still requires having a Godot module. So, at the moment, we don't _need_ to move the .NET specific loader into the GDExtension, since we'll have a module anyway.

> I think that we should take this opportunity to move away from the path based GDExtension identification concept to a more generic solution. For each GDExtension we could have a GDExtensionDescriptor (which is RefCounted), that uniquely describes that GDExtension.

Hm. I'm not crazy about this idea. Almost all GDExtensions will have a path they can point to, even if that path isn't a `.gdextension` file. The one exception to this is LibGodot, and I think making up fake paths for that one case is fine. We could even use a special prefix (`libgodot://` or something) so that they can't be confused for real files.

> I think this solution would provide us with a future-proof architecture, where GDExtensions can also provide custom GDExtension loader implementations.

Rather than "future-proofing", I'd prefer to work on the architecture internally first, and then when we're more confident about it, figure out how to expose it externally. At that point we'll know more about it, and be able to do a better job designing its API.

Reply:
> The .NET GDExtension implementation could implement the .NET specific loader in itself

Like David said .NET will use a new `dotnet` module. The `dotnet` module will implement a `GDExtensionLoader` that handles loading .NET GDExtensions using hostfxr or NativeAOT. This is already implemented using this PR, but I wanted to get this PR merged first.

> I think that we should take this opportunity to move away from the path based GDExtension identification concept to a more generic solution.

I agree with David, I think using paths (doesn't have to be a real path) works well enough to identify the extensions and then each `GDExtensionLoader` can handle it however it wants. For example, the .NET `GDExtensionLoader` uses the path to the DLL but it may use different configuration depending on whether we are loading the assembly with hostfxr or NativeAOT.

> For each GDExtension we could have a GDExtensionDescriptor (which is RefCounted), that uniquely describes that GDExtension.

I think this can be handled by `GDExtensionLoader` already. I want to clarify that `GDExtensionLoader` is not a factory, there's one instance for each extension, so each loader instance can have data specific to the extension and that sounds similar to what you define as a `GDExtensionDescriptor`.

Reply:
> > I think it would be beneficial to allow the definition of GDExtensionLoaders inside GDExtensions
> 
> This is a very interesting idea, and I think it's something we should allow eventually! But I don't think we should rush into it.
> 
> The way the current PR is written has no effect on Godot's external API - the `GDExtensionLoader` here is a completely internal concept. This allows us to try it out, and iterate on it, without the risk of accidentally committing to a broken API that we need to maintain compatibility for.

I think if we mark the API experimental that should be clear to anyone that it can change (or even disappear without notice.)

> 
> > I think that we should take this opportunity to move away from the path based GDExtension identification concept to a more generic solution. For each GDExtension we could have a GDExtensionDescriptor (which is RefCounted), that uniquely describes that GDExtension.
> 
> Hm. I'm not crazy about this idea. Almost all GDExtensions will have a path they can point to, even if that path isn't a `.gdextension` file. The one exception to this is LibGodot, and I think making up fake paths for that one case is fine. We could even use a special prefix (`libgodot://` or something) so that they can't be confused for real files.

I can accept that: introducing a custom URL scheme like ``libgodot://`` seems like a good compromise. We will rework the libgodot patch accordingly.

Reply:
> > The .NET GDExtension implementation could implement the .NET specific loader in itself
> 
> Like David said .NET will use a new `dotnet` module. The `dotnet` module will implement a `GDExtensionLoader` that handles loading .NET GDExtensions using hostfxr or NativeAOT. This is already implemented using this PR, but I wanted to get this PR merged first.
> 

Could you please explain the advantages of this design?

One big disadvantage I see is that we still need to provide separate .NET and Standard builds of the engine, which is something that AFAIK the release team wanted to get rid of for years now.

It also adds a strong coupling of the .NET binding to engine versions. If a GDExtension could implement loaders on its own, then the .NET support could be added to any project as an asset, and could be potentially updated more frequently than the core engine, or have multiple asset versions with different .NET version support ... etc..).  

In addition to the above, as I mentioned, if GDExtensions could define their own loaders, other languages could add the same level of support to Godot as .NET, without modifying the core engine.

I also worry, that the current implementation would make it harder to support the libgodot use case, where the .NET app initializes Godot and then also initializes the .NET GDExtension bindings so the .NET Godot API can be used to control Godot from the hosting .NET app.

What do you think?

PS: Since this PR is also targeting 4.4, I think we have time to ensure we don't mess up the exposed GDExtensionLoader interface. We could also just involve the people working on the different language bindings (e.g. Java or Python) to see if they have some additional requirements.

Reply:
> One big disadvantage I see is that we still need to provide separate .NET and Standard builds of the engine, which is something that AFAIK the release team wanted to get rid of for years now.

No, the module will be enabled in Standard builds of the engine, like the GDScript module.

> It also adds a strong coupling of the .NET binding to engine versions. If a GDExtension could implement loaders on its own, then the .NET support could be added to any project as an asset

This is precisely what we want to avoid, we don't want C# users having to add an extension to every project to get C# support. The goal is for C# to feel as integrated as GDScript, not less.

> and could be potentially updated more frequently than the core engine

The module only contains the assembly loading which I don't think would change often. For the editor integration (what we call `GodotTools` in the current system) we'll have a separate GDExtension that is downloaded on-demand and can be updated separately. The bindings package (what we call `GodotSharp` in the current system) is already a separate package and this one depends on the user extension since it's a direct dependency.

> if GDExtensions could define their own loaders, other languages could add the same level of support to Godot as .NET

That's the goal of the `GDExtensionLoader` class. When it's exposed to GDExtension / scripting, others will be able to implement custom loaders or use existing exposed loaders with `GDExtensionManager::load_extension_with_loader`.

> I also worry, that the current implementation would make it harder to support the libgodot use case, where the .NET app initializes Godot and then also initializes the .NET GDExtension bindings so the .NET Godot API can be used to control Godot from the hosting .NET app.

I'm not very familiar with LibGodot but I don't see how the current implementation makes it harder to support. In this case you wouldn't use the .NET GDExtensionLoader because the runtime is already initialized and the assembly is loaded, and the entry-point to initialize it (and the bindings) is the same as in any other language (e.g.: `godot-cpp`).

> Since this PR is also targeting 4.4, I think we have time to ensure we don't mess up the exposed GDExtensionLoader interface. We could also just involve the people working on the different language bindings (e.g. Java or Python) to see if they have some additional requirements.

Yes, that's why we're not exposing the class yet, we want other language bindings authors to provide feedback. But I don't want to delay this PR with additional improvements that would be irrelevant to my work on the `dotnet` module which is already sufficiently covered by the current implementation.

Reply:
@raulsntos Thank you for the detailed explanation. Could I look at the new dotnet module code that implements the GDExtensionLoader (I could not find it in your public Godot fork.)?

Reply:
It's not public yet, there's still a couple of things that I want to cleanup. I can follow-up on [chat.godotengine.org](https://chat.godotengine.org) and let you know when it's ready.